### PR TITLE
rebuild when the build type changes

### LINKED ISSUE #3788: flutter run doesn't know that a user switched from debug to release
Consider this series of steps:

* flutter run
* flutter run --release

Expected: the second run should be in release mode.

Actual: the second run is still using debug mode.

From the logs:

```
  4 ms • Running build command.
   0 ms • APK up to date; skipping build step.
   4 ms • Stopping app 'app.apk' on Nexus 5.
```

Workaround: `rm -rf build` before you toggle `--release` (or, before you switch between modes).

Could we check if the requested mode isn't the previous mode, and use different assets in build?

@jason-simmons suggested a layout something like `build/debug` and `build/release` etc.

Thanks!

### PR DESCRIPTION
**Pull request by devoncarew**
Redux of https://github.com/jason-simmons/flutter/pull/1, rebuild the apk when the build type (or target device type) changes. Fix https://github.com/flutter/flutter/issues/3788.

### CODE REVIEW

**Comment by abarth** [comment]
LGTM

**Comment by abarth** [comment]
Does this handle switching between locally built engines and cloud engines?

**Comment by devoncarew** [comment]
> Does this handle switching between locally built engines and cloud engines?

No, it doesn't know about local engine builds; I'll add support for that in this PR.

**Comment by devoncarew** [comment]
PTAL, updated to take local engine builds into account.

File: packages/flutter_tools/lib/src/commands/build_apk.dart [hunk 1]
```diff
@@ -547,3 +564,24 @@ Future<int> buildApk(
...
+  Map<String, dynamic> meta = _readBuildMeta(buildDirectoryPath);
+  meta[key] = value;
+  File buildMetaFile = new File(path.join(buildDirectoryPath, 'build_meta.json'));
+  buildMetaFile.writeAsStringSync(toPrettyJson(meta));
+}
+
+String _getTargetBuildTypeToken(TargetPlatform platform, BuildMode buildMode) {
+  String buildType = getNameForTargetPlatform(platform) + '-' + getModeName(buildMode);
+  if (tools.isLocalEngine)
+    buildType += ' [${tools.engineBuildPath}]';
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/build_apk.dart]: this should include the binary's last modified time

Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/build_apk.dart]: makes sense; done

**Comment by devoncarew** [comment]
landing...

**Pull request merged**